git.md

```diff
@@ -0,0 +1,536 @@
+"solution" subdirectory to my repository. Let’s simulate that
+process. Go into the original test1 directory, and make the solution
+subdirectory and create two files in it:
+
+```bash
+cd ../test1
+mkdir solution
+cd solution
+cp ../hello.c .
+echo 'hello:' > Makefile
+```
+
+Type `ll` to see that two files (`Makefile` and `hello.c`) have been created
+in the `solution` directory. (`hello.c` was copied from the parent
+directory, and Makefile was created directly on the command line using
+the echo command. BTW, the Makefile contains a single line, `hello:`.
+Can you see why this is a legitimate Makefile?)
+Now, move out of the solution directory, and `git add` and `git commit` the
+solution directory:
+
+```bash
+cd ..
+git add solution
+git commit -m "added solution"
+```
+
+Note that `git add solution` stages all files in the directory.
+
+## Creating tags from the command line
+
+There are two steps to go when submitting your homework in this class.
+First, you have to "tag" the commit
+you want to hand in before "pushing" it to the remote server.
+When you use git tag <tagname>, Git will create a tag at the current revision
+but will not prompt you for an annotation. It will be tagged without a message.
+When you use git tag -a -m <msg> <tagname>,
+Git will tag the commit and annotate it with the provided message.
+For example, I finish my hw0 and tend to submit
+
+```bash
+git tag -a -m "hw0 submission" handin
+```
+
+Git will tag the commit and annotate it with the provided message
+with message "hw0 submission" and tag it as "handin".
+
+To delete a tag on your local repository, you can use git tag -d <tagname>
+
+```bash
+git tag -d handin
+```
+
+It can remove the handin tag just made in your local machine.
+
+## Pushing commits to a remote repository
+
+Assuming you didn't delete the handin tag, you are ready to push
+your "hw0 submission" to the remote server.
+The `git push` command takes two arguments:
+    A remote name, for example, origin
+    A branch name, for example, main
+
+```bash
+git push origin main
+```
+
+will do the job, and now your submission is on the remote server.
+If you want to re-submit after pushing, you have to delete the
+old tag in the remote server first:
+
+```bash
+git push --delete origin handin
+```
+
+and then remove the local handin tag before tagging the new commit for another push.
+
+## Pulling changes from a remote repository
+
+Once you hear that the lab solution is available, you will want to
+retrieve it and take a look. You do that by "pulling" the changes in
+my repository into your repository. Let’s pull the changes we just
+made in `test1` into `test2`:
+
+```bash
+cd ../test2/
+git pull
+```
+
+The `git pull` command looks at the original repository that you
+cloned from, fetches all the changes made since the cloning, and
+merges the changes into the current repository. You now have the
+solution right in your repository.
+
```

Comment:
```suggestion
## Branches
With group assignments, we recommend that you push work to branches first, and then merge back into main once your group members have reviewed the code. As an example, suppose that you are working on a part of the assignment, you can create a branch separate from main by doing the following:

`$ git checkout main`
`$ git checkout -b <branch-name>`

You can then commit your changes, and push to the branch by doing the following:

`$ git push origin <branch-name>`

This will allow multiple members of the team to work on separate features in parallel. When the feature you are working on is complete, you may then create a pull request to allow your team members to review the code, and finally merge the changes back into master. You can read more about using branches and pull requests from [GitHub’s own documentation](https://help.github.com/articles/proposing-changes-to-your-work-with-pull-requests/).
```

Reply:
this didn't format well....